scripts/verify_commits.sh

```diff
@@ -0,0 +1,117 @@
+
+    local -a upstream_commits
+    mapfile -t upstream_commits < <(echo "${log}" | grep 'Upstream-commit:' | awk '{print $2}')
+
+    if (( ${#upstream_repos[@]} < 1 && ${#upstream_commits[@]} < 1 )); then
+        # no upstream commit referenced
+        return 0
+    fi
+
+    if (( ${#upstream_repos[@]} > 1 )); then
+        err "downstream staging commit ${downstream_commit} references more than one upstream repo:"
+        err "${upstream_repos[@]}"
+        err "staging commits may only reference a single upstream repo"
+        return 1
+    fi
+
+    if (( ${#upstream_commits[@]} > 1 )); then
+        err "downstream staging commit ${downstream_commit} references more than one upstream commit:"
+        err "${upstream_commits[@]}"
+        err "staging commits may only reference a single upstream commit"
+        return 1
+    fi
+
+    if git branch -r --contains "${upstream_commits[0]}" | grep -vq "${upstream_repos[0]}/.*"; then
+        err "downstream staging commit ${downstream_commit} references upstream commit ${upstream_commits[0]} that doesn't exist in ${upstream_repos[0]}"
+        err "staging commits must reference a repository containing the given upstream commit"
+        return 1
+    fi
+
+    echo "${upstream_repos[0]}"
+    echo "${upstream_commits[0]}"
+}
+
+function fetch_remotes() {
+    local -a remotes
+    remotes=(
+        'api'
+        'operator-registry'
+        'operator-lifecycle-manager'
+    )
+    for r in "${remotes[@]}"; do
+        local url
+        url="https://github.com/operator-framework/${r}.git"
+
+        git remote add "${r}" "${url}" || git remote set-url "${r}" "${url}"
+        git fetch "${r}" master
+    done
+}
+
+function main() {
+    fetch_remotes || { err "failed to fetch remotes" &&  exit 1; }
+
+    local target_branch="${1:-master}"
+
+    # get all commits we're introducing into the target branch
+    local -a new_commits
+    mapfile -t new_commits < <(git rev-list --no-merges HEAD "^${target_branch}")
+    info "detected ${#new_commits[@]} commits to verify"
+
+    local -a sr
+    for commit in "${new_commits[@]}"; do
+        info "verifying ${commit:0:7}..."
+        info "$(git log -n 1 "${commit}")"
+
+        mapfile -t sr < <(upstream_ref "${commit}" || exit 1)
+        if (( ${#sr[@]} < 2 )); then # the ref contains a tuple if the values were properly parsed from the commit message
+            # couldn't find upstream cherry-pick reference in the commit message
+            # assume it's downstream-only commit
+            info "${commit:0:7} does not reference upstream commit, verifying as downstream-only"
+        else
+            # found upstream cherry-pick reference in commit message
+            # verify as an upstream sync
+            info "${commit:0:7} references upstream commit (${sr[*]}), verifying as upstream sync"
+            verify_upstream_sync "${sr[0]}" "${commit}" || exit 1
+        fi
```

Comment:
Hmm it's possible I'm misreading this but this seems like it's avoiding the edge case where we're pulling in an upstream commit, but someone forgot to populate that upstream commit metadata into the pulled down commit?

Reply:
Ah, yeah I glossed over the commented text - it seems like we'd want to enforce this behavior as a stop gap until automation has been introduced. Would it be more reasonable to assume that anything that touches the staging tree _needs_ to be treated as an upstream commit? I still need to dive  into that upstream_ref function so I may be missing some context here.